x/trigger/types/trigger.go

```diff
@@ -0,0 +1,182 @@
+package types
+
+import (
+	fmt "fmt"
+	"strings"
+	time "time"
+
+	proto "github.com/gogo/protobuf/proto"
+
+	abci "github.com/tendermint/tendermint/abci/types"
+
+	codectypes "github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdktx "github.com/cosmos/cosmos-sdk/types/tx"
+)
+
+type TriggerID = uint64
+
+const (
+	BlockHeightPrefix = "block-height"
+	BlockTimePrefix   = "block-time"
+)
+
+type TriggerEventI interface {
+	proto.Message
+	GetEventPrefix() string
+	Validate() error
+	ValidateContext(ctx sdk.Context) error
+}
+
+var _ TriggerEventI = &TransactionEvent{}
+var _ TriggerEventI = &BlockHeightEvent{}
+var _ TriggerEventI = &BlockTimeEvent{}
+
+// NewTrigger creates a new trigger.
+func NewTrigger(id TriggerID, owner string, event *codectypes.Any, action []*codectypes.Any) Trigger {
+	return Trigger{
+		id,
+		owner,
+		event,
+		action,
+	}
+}
+
+// NewQueuedTrigger creates a new trigger for queueing.
+func NewQueuedTrigger(trigger Trigger, blockTime time.Time, blockHeight uint64) QueuedTrigger {
+	return QueuedTrigger{
+		Time:        blockTime,
+		BlockHeight: blockHeight,
+		Trigger:     trigger,
+	}
+}
+
+// Equals checks if two TransactionEvents have the same type and attributes.
+func (e TransactionEvent) Equals(other abci.Event) bool {
+	if e.Name != other.GetType() {
+		return false
+	}
+
+	for _, attr := range e.Attributes {
+		hasAttribute := false
+
+		for _, otherAttr := range other.Attributes {
+			if attr.Equals(otherAttr) {
+				hasAttribute = true
+				break
+			}
+		}
+
+		if !hasAttribute {
+			return false
+		}
+	}
+
+	return true
+}
```

Comment:
Nit: This isn't really an `Equals` function. Primarily, it only checks that `e`'s attributes exist in the `other`s. But it doesn't take into account a situation where `other` has extra attributes or `e` has a duplicated attribute entry but `other` only has that entry once.